### DISCUSSION TITLE
How to deserialize json with "random" property name?

### DISCUSSION THREAD

**Discussion by Symbai** [dotnet/runtime#77337] [Q&A] [answered]
I have to deserialize json where one property name can be anything. How can I do this?

Example:
```JSON
{
  "57690": { ... }
}
```

Where 57690 can also be any other number. But it is the only property at this level.

```C#
    public class Root
    {
        public SubClass _Random_Number_Here_ { get; set; } // <--- How can I make this 'Universal' to 57690 but also 12345 etc
    }

var bla = JsonSerializer.Deserialize<Root>(...);
```

**Comment by TonyValenti** [comment]
I would deserialize the object as a dictionary and then access the value that way.

**Reply by TonyValenti** [reply]
Alternatively, you can also deserialize it as a JsonDocument, find the node, and then deserialize that value. 

Also - you should tell the developers of that api that their json model sucks.

**Reply by Symbai** [reply]
I dont want to access values via dictionary because I bind data from it and it has many nested levels. How do I find the node and deserialize the value? All code examples I found are accessing the child nodes via property name, which I ofc don't know.

**Comment by samsosa** [comment, marked as answer]
Maybe you can use the [[JsonExtensionData](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonextensiondataattribute?view=net-6.0)] attribute?

**Reply by Symbai** [reply]
This is what I did and changed property type to [JsonObject](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.nodes.jsonobject?f1url=%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(System.Text.Json.Nodes.JsonObject)%3Bk(DevLang-csharp)%26rd%3Dtrue&view=net-6.0). Unfortunately I have no idea how I can get the value from JsonObject without knowing the property name alias "57690". There is no property or method in JsonObject to retrieve its name / key. There is no property to retrieve its value without knowing the keys name.

**Reply by samsosa** [reply]
Hmm, I would approach the problem as follows:

```csharp
    class Program
    {
        static async Task Main(string[] args)
        {
            Root startInstance = JsonSerializer.Deserialize<Root>("{ \"57690\": { \"SomeText\": \"Hello World\" } }");

            String text = startInstance._Random_Number_Here_.SomeText;
            startInstance._Random_Number_Here_ = new Subclass { SomeText = "Another World" };

            String finalJson = JsonSerializer.Serialize(startInstance);

            Console.WriteLine(finalJson);

            Console.ReadKey();
        }

        public class Root
        {
            [JsonExtensionData]
            public Dictionary<string, JsonElement> ExtensionData { get; set; }

            [JsonIgnore]
            public Subclass _Random_Number_Here_
            {
                get
                {
                    if (ExtensionData?.Count == 1)
                    {
                        foreach (var el in ExtensionData)
                        {
                            return el.Value.Deserialize<Subclass>();
                        }
                    }

                    return default;
                }

                set
                {
                    if (ExtensionData?.Count == 1)
                    {
                        foreach (var el in ExtensionData)
                        {
                            ExtensionData[el.Key] = JsonSerializer.SerializeToElement(value);
                            break;
                        }
                    }
                }
            }
        }

        public class Subclass
        {
            public String SomeText { get; set; }
        }
    }
```

**Reply by Symbai** [reply]
Thank you, that probably works
